make arraymancer work with devel

https://pipelines.actions.githubusercontent.com/ZRinn1OrR0LWxU3iWy1StaQcZRN2kXW9lHWwDJ3esfasrmfdRn/_apis/pipelines/1/runs/18644/signedlogcontent/8?urlExpires=2021-04-18T15%3A29%3A02.1515769Z&urlSigningMethod=HMACV1&urlSignature=ES5o3LHv4%2Bky4FEHjiSAzReQNwqIZDrkd2X%2B6tFYxcU%3D


```
2021-04-18T13:40:09.5027645Z                  from /usr/lib/gcc/x86_64-linux-gnu/7/include/x86intrin.h:48,
2021-04-18T13:40:09.5028566Z                  from /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:11:
2021-04-18T13:40:09.5031638Z /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c: In function ‘gebb_ukernel_int64_x86_AVX512_tensorZtest95operators95blas_37425’:
2021-04-18T13:40:09.5034234Z /usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:282:1: error: inlining failed in call to always_inline ‘_mm512_setzero_si512’: target specific option mismatch
2021-04-18T13:40:09.5035235Z  _mm512_setzero_si512 (void)
2021-04-18T13:40:09.5035871Z  ^~~~~~~~~~~~~~~~~~~~
2021-04-18T13:40:09.5036752Z /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:23985:9: note: called from here
2021-04-18T13:40:09.5037563Z   AB13_1 = _mm512_setzero_si512();
2021-04-18T13:40:09.5038016Z   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
2021-04-18T13:40:09.5039014Z In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:45:0,
2021-04-18T13:40:09.5040103Z                  from /usr/lib/gcc/x86_64-linux-gnu/7/include/x86intrin.h:48,
2021-04-18T13:40:09.5041044Z                  from /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:11:
2021-04-18T13:40:09.5042887Z /usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:282:1: error: inlining failed in call to always_inline ‘_mm512_setzero_si512’: target specific option mismatch
2021-04-18T13:40:09.5043892Z  _mm512_setzero_si512 (void)
2021-04-18T13:40:09.5044319Z  ^~~~~~~~~~~~~~~~~~~~
2021-04-18T13:40:09.5045218Z /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:23982:9: note: called from here
2021-04-18T13:40:09.5046020Z   AB13_0 = _mm512_setzero_si512();
2021-04-18T13:40:09.5046491Z   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
2021-04-18T13:40:09.5047498Z In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:45:0,
2021-04-18T13:40:09.5048610Z                  from /usr/lib/gcc/x86_64-linux-gnu/7/include/x86intrin.h:48,
2021-04-18T13:40:09.5049536Z                  from /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:11:
2021-04-18T13:40:09.5051321Z /usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h:282:1: error: inlining failed in call to always_inline ‘_mm512_setzero_si512’: target specific option mismatch
2021-04-18T13:40:09.5052304Z  _mm512_setzero_si512 (void)
2021-04-18T13:40:09.5052743Z  ^~~~~~~~~~~~~~~~~~~~
2021-04-18T13:40:09.5053612Z /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:23979:9: note: called from here
2021-04-18T13:40:09.5054412Z   AB12_1 = _mm512_setzero_si512();
2021-04-18T13:40:09.5054871Z   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
2021-04-18T13:40:09.5055686Z compilation terminated due to -fmax-errors=3.
2021-04-18T13:40:09.5058447Z Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/home/runner/work/Nim/Nim/lib -I/home/runner/work/Nim/Nim/pkgstemp/arraymancer/tests -o /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c.o /home/runner/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c' failed with exit code: 1
```


and https://github.com/nim-lang/Nim/runs/2374356299

--- a/README.md
+++ b/README.md
@@ -31,7 +31,7 @@ Reminder of supported compilation flags:
   See [nimblas](https://github.com/unicredit/nimblas) for further configuration.
 - `-d:cuda`: Build with Cuda support
 - `-d:cudnn`: Build with CuDNN support, implies `cuda`.
-  - `-d:avx512`: Build with AVX512 support by supplying the
+- `-d:avx512`: Build with AVX512 support by supplying the
   `-mavx512dq` flag to gcc / clang. Without this flag the
   resulting binary does not use AVX512 even on CPUs that support
   it. Handing this flag however, makes the binary incompatible with

--- a/README.md
+++ b/README.md
@@ -31,6 +31,12 @@ Reminder of supported compilation flags:
   See [nimblas](https://github.com/unicredit/nimblas) for further configuration.
 - `-d:cuda`: Build with Cuda support
 - `-d:cudnn`: Build with CuDNN support, implies `cuda`.
+  - `-d:avx512`: Build with AVX512 support by supplying the
+  `-mavx512dq` flag to gcc / clang. Without this flag the
+  resulting binary does not use AVX512 even on CPUs that support
+  it. Handing this flag however, makes the binary incompatible with
+  CPUs that do *not* support it. See the comments in #505 for a
+  discussion (from `v0.7.9`).
 - You might want to tune library paths in [nim.cfg](nim.cfg) after installation for OpenBLAS, MKL and Cuda compilation.
   The current defaults should work on Mac and Linux.
 

--- a/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim
+++ b/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim
@@ -228,25 +228,43 @@ proc gemm_strided*[T: SomeNumber and not(uint32|uint64|uint|int)](
         const ukernel = cpu_features.x86_ukernel(T, false)
         apply(ukernel)
 
-    when defined(i386) or defined(amd64):
-      when T is float32:
-        if hasAvx512f(): dispatch(x86_AVX512)
-        elif hasFma3():  dispatch(x86_AVX_FMA)
-        elif hasAvx():   dispatch(x86_AVX)
-        elif hasSse():   dispatch(x86_SSE)
-      elif T is float64:
-        if hasAvx512f(): dispatch(x86_AVX512)
-        elif hasFma3():  dispatch(x86_AVX_FMA)
-        elif hasAvx():   dispatch(x86_AVX)
-        elif hasSse2():  dispatch(x86_SSE2)
-      elif T is int32:
-        if hasAvx512f(): dispatch(x86_AVX512)
-        elif hasAvx2():  dispatch(x86_AVX2)
-        elif hasSse41(): dispatch(x86_SSE4_1)
-        elif hasSse2():  dispatch(x86_SSE2)
-      elif T is int64:
-        if hasAvx512f(): dispatch(x86_AVX512)
-        elif hasSse2():  dispatch(x86_SSE2)
+    # for clarity split AVX512 compilation fully from regular
+    when defined(avx512):
+      when defined(i386) or defined(amd64):
+        when T is float32:
+          if hasAvx512f(): dispatch(x86_AVX512)
+          elif hasFma3():  dispatch(x86_AVX_FMA)
+          elif hasAvx():   dispatch(x86_AVX)
+          elif hasSse():   dispatch(x86_SSE)
+        elif T is float64:
+          if hasAvx512f(): dispatch(x86_AVX512)
+          elif hasFma3():  dispatch(x86_AVX_FMA)
+          elif hasAvx():   dispatch(x86_AVX)
+          elif hasSse2():  dispatch(x86_SSE2)
+        elif T is int32:
+          if hasAvx512f(): dispatch(x86_AVX512)
+          elif hasAvx2():  dispatch(x86_AVX2)
+          elif hasSse41(): dispatch(x86_SSE4_1)
+          elif hasSse2():  dispatch(x86_SSE2)
+        elif T is int64:
+          if hasAvx512f(): dispatch(x86_AVX512)
+          elif hasSse2():  dispatch(x86_SSE2)
+    else:
+      when defined(i386) or defined(amd64):
+        when T is float32:
+          if hasFma3():    dispatch(x86_AVX_FMA)
+          elif hasAvx():   dispatch(x86_AVX)
+          elif hasSse():   dispatch(x86_SSE)
+        elif T is float64:
+          if hasFma3():    dispatch(x86_AVX_FMA)
+          elif hasAvx():   dispatch(x86_AVX)
+          elif hasSse2():  dispatch(x86_SSE2)
+        elif T is int32:
+          if hasAvx2():    dispatch(x86_AVX2)
+          elif hasSse41(): dispatch(x86_SSE4_1)
+          elif hasSse2():  dispatch(x86_SSE2)
+        elif T is int64:
+          if hasSse2():  dispatch(x86_SSE2)
     dispatch(x86_Generic)
 
 proc gemm_strided*[T: uint32|uint64|uint|int](

--- a/src/arraymancer/laser/primitives/matrix_multiplication/gemm_ukernel_avx512.nim
+++ b/src/arraymancer/laser/primitives/matrix_multiplication/gemm_ukernel_avx512.nim
@@ -9,6 +9,12 @@ import
 
 x86only()
 
+## For the C codegen of AVX512 instructions to be valid, we need the following flag:
+when defined(avx512) and (defined(gcc) or defined(clang)):
+  {.passC: "-mavx512dq".}
+## See: https://stackoverflow.com/a/63711952
+## for a script to find the required compilation flags for specific SIMD functions.
+
 ukernel_generator(
     x86_AVX512,
     typ = float32,

